Add command to generate behaviour

JoeChapman opened the pull request:
This PR adds `behaviour` to the binaries to create a new hof behaviour project.

lennym commented:
What do you expect this command to actually do? I can't work out what "create behaviour" is intended to mean.

lennym commented:
git submodules? For realsies?

lennym commented on README.md:
```diff
@@ -37,3 +37,14 @@ $ hof app <name>
+### Options
+
+* `dir` - the root directory of your application - Default `.`
+* `name` - the name of your first hof behaviour - Default `default-behaviour`
```
A default value that won't ever survive is probably a bad default and shouldn't exist.

lennym commented on lib/cli/behaviour/defaults.js:
```diff
@@ -0,0 +1,10 @@
+  dir: process.cwd(),
+  force: false,
+  name: 'default-behaviour',
+  secret: crypto.randomBytes(64).toString('hex')
```
What is `secret` for in this context? I suspect a relic of copy-pasting the `init` code.

lennym commented on lib/cli/behaviour/tasks/add-gitignore.js:
```diff
@@ -0,0 +1,10 @@
+
+module.exports = function gitignore(settings) {
+  return append(path.resolve(settings.dir, '.gitignore'), ignore.join('\n') + '\n');
+};
```
This is appending some nothing to the `.gitignore` and can probably be completely removed.

If it's going to continue to exist it should probably find some commonality with the `init` version.

lennym commented on lib/cli/behaviour/tasks/check-empty.js:
```diff
@@ -0,0 +1,25 @@
+Use '--force' flag to force overwrite existing contents`);
+      }
+    });
+};
```
This file is a straight up copy of the `init` version, and should probably be shared. The error message content should maybe be rewritten too.

lennym commented on lib/cli/behaviour/tasks/copy-files.js:
```diff
@@ -0,0 +1,8 @@
+
+module.exports = function copy(settings) {
+  return cp(path.resolve(__dirname, '../template'), settings.dir, settings);
+};
```
This can be re-used from the `init` version.

lennym commented on lib/cli/behaviour/tasks/update-package.js:
```diff
@@ -0,0 +1,21 @@
+    .catch(() => {
+      throw new Error(`Cannot update contents at ${settings.dir}/package.json`);
+    });
+};
```
This would be better done just by adding a placeholder to the template package.json.

If a package.json already exists in the directory then I think it would be better to preserve the name that's already there.

lennym commented:
This seems overly complex for something that is ultimately just going to create a barebones package.json and an index.js with `module.exports = superclass => class extends superclass {}`

I'm also not sure that the template files have been added correctly.

JoeChapman commented:
The submodule is an accident. Sorry, I didn't really take enough care over this but I wanted to get something up to discuss. 
The submodule must have been created from copying over `.git` from `hof-behaviour-summary-page, which I used to create the template files.

JoeChapman commented:
Removed the nested `.git` folder that was accidentally copied over which created the unwanted submodule. 
Moved shared tasks to `lib/tasks/` and removed the tasks form behaviours that aren't required.

JoeChapman commented:
@lennym, I made changes

lennym commented on lib/cli/behaviour/defaults.js:
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+  dir: process.cwd(),
+  force: false,
+  name: 'default-behaviour',
```
I don't think there's value in this default. I'd make the default be the name of the folder in which the command is being run.

The most common workflow will be:

* create new git repo
* clone empty git repo
* `cd <repo name>`
* `hof behaviour .`

So if the default name is the directory name, then it will match the repo name, which I think is the best approach and mostly likely desired behaviour.

lennym commented:
This still seems to have a submodule at `lib/cli/behaviour/template` rather than the actual files themselves.

lennym commented on lib/cli/behaviour/template/.eslintignore:
```diff
@@ -0,0 +1,37 @@
+.npm
+
+# Optional REPL history
+.node_repl_history
```
Does this need to exist?

Other behaviours have minimal .eslintignore files - https://github.com/UKHomeOfficeForms/hof-behaviour-session/blob/master/.eslintignore

Edit: I've just noticed this is copied from hof-behaviour-summary-page, which I added. I think that was a mistake.

JoeChapman replied:
You are totally right, why haven't I added a `package.json`??? worst PR ever!

JoeChapman replied:
well this is embarrassing, I did add  `lib/cli/behaviour/lib` and  `lib/cli/behaviour/index.js` but they did not survive the process of removing the submodule.

JoeChapman commented:
I imagined you 'd used the default (it looked like a default bunch of .dotfiles) .dotfiles, as generated by Express, because we don't want to dictate too much to the user.

JoeChapman commented:
It's just possible that in attempting to add a behaviour generator, we've discovered it's not necessary

easternbloc commented:
What we doing with this PR? @JoeChapman do you still think it's relevant?

sulthan-ahmed commented:
This PR is 103 days 17h 28m old, as @easternbloc mentioned, do we know what we want to do with this one @JoeChapman  ?
